feat(bookings): implement Booking UID filter for bookings data table.

akarsh-jain-790 opened the pull request:
## What does this PR do?
Adds booking UID filter to the bookings listing view, allowing users to search bookings by their unique identifier.

Fixes: #21357  
Fixes [CAL-5780](https://linear.app/calcom/issue/CAL-5780/add-booking-uid-to-the-filter-options-on-bookings)

## Visual Demo

<img width="1244" alt="Screenshot 2025-05-18 at 1 00 03 PM" src="https://github.com/user-attachments/assets/f8291df1-c83d-435b-ad3d-e3c16626b505" />

<table>
  <tr>
    <td align="center">
      <strong>Before</strong><br>
      <img src="https://github.com/user-attachments/assets/7db659b7-3307-4bd6-95a2-d089dd2dc3ef" width="232" />
    </td>
    <td align="center">
      <strong>After</strong><br>
      <img src="https://github.com/user-attachments/assets/e09591b1-06b7-4c70-8a66-3707a0c22d47" width="253" />
    </td>
  </tr>
</table>

## Mandatory Tasks
- [x] I have self-reviewed the code
- [N/A] I have updated the developer docs
- [x] I confirm automated tests are in place

## Bug Fixes/Features
- New filter appears in booking list filter dropdown
- Works consistently across all booking status views
- Case-insensitive search implemented

## Mandatory Tasks (DO NOT REMOVE)

- [X] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [X] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## Summary
- Added booking UID filter to bookings listing
- Supports both exact matches and partial text search
- Maintains all existing security and permission checks
    
<!-- This is an auto-generated description by cubic. -->
---

## Summary by cubic
Added a booking UID filter to the bookings table, letting users search bookings by their unique identifier.

- **New Features**
  - Booking UID filter appears in the filter dropdown.
  - Supports exact and partial text search.
  - Works across all booking status views.

<!-- End of auto-generated description by cubic. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=21380) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=21380) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=21380) it.</sub>

akarsh-jain-790 committed:
feat(bookings): implement Booking UID filter for bookings data table.

Fixes: #21357

akarsh-jain-790 commented:
Hi @eunjae-lee, could you please review this PR? I added a Booking UID filter to the bookings data table, and since you’ve worked on similar features, I’d value your feedback. This is my first PR, so any guidance is appreciated! Let me know if you need more context or testing details.

TusharBhatt1 reviewed (commented):
LGTM

eunjae-lee commented:
> Hi @eunjae-lee, could you please review this PR? I added a Booking UID filter to the bookings data table, and since you’ve worked on similar features, I’d value your feedback. This is my first PR, so any guidance is appreciated! Let me know if you need more context or testing details.

Hey @akarsh-jain-790 I already have an on-going PR on the same topic. I guess you may have referred to it. Anyway I'm happy to close that PR and assist you to continue to work here.

If you look at the second commit https://github.com/calcom/cal.com/pull/21358/commits, I added the option to the text filter so that it allows only "equals" operator, as booking UID filter doesn't make sense with partial match. Could you apply the change here?

Also, now that we're enforcing "equals" operator only, the input schema should also be

```diff
-bookingUid: z.union([z.string(), ZTextFilterValue]).optional(),
+bookingUid: z.string().optional(),
```

and update the logics on the backend side accordingly.

Does it make sense? Feel free to reply to me if anything is unclear to you!

akarsh-jain-790 commented:
Hey @eunjae-lee Thanks for pointing out your PR. I wasn’t aware of it as I was referring to a different PR for another filter implementation. I've implemented both changes you mentioned: the text filter is now restricted to only the `equals` operator, and I've updated the schema to use `z.string().optional()` instead of the union type.

eunjae-lee commented:
@akarsh-jain-790 thanks! lovely!
